File: mocks/oppgave-mock/src/main/java/no/nav/pensjon/vtp/mocks/oppgave/OppgaveMockImpl.java
package no.nav.pensjon.vtp.mocks.oppgave;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.swagger.annotations.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.validation.Valid;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.HttpHeaders;
import javax.ws.rs.core.MultivaluedMap;
import javax.ws.rs.core.UriInfo;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;

@RestController
@Api(tags = "Oppgave Mock")
@RequestMapping(value = "oppgave/api/v1/oppgaver",
        produces = org.springframework.http.MediaType.APPLICATION_JSON_VALUE,
        consumes = org.springframework.http.MediaType.APPLICATION_JSON_VALUE)
public class OppgaveMockImpl {

    private static final Logger LOG = LoggerFactory.getLogger(OppgaveMockImpl.class);

    private static final Map<Long, ObjectNode> oppgaver = new ConcurrentHashMap<>();

    @PostMapping
    @ApiOperation(value = "Opprett oppgave", response = OppgaveJson.class)
    @ApiImplicitParams({
            @ApiImplicitParam(name = "X-Correlation-ID", required = true, dataType = "string", paramType = "header"),
            @ApiImplicitParam(name = "Authorization", required = true, dataType = "string", paramType = "header")

    })
    public ResponseEntity opprettOppgave(
            @Valid @ApiParam(value = "Oppgaven som opprettes", required = true) ObjectNode oppgave,
            @Context HttpHeaders httpHeaders) {
        Optional<ResponseEntity> validert = validerIkkeFunksjonelt(httpHeaders);
        if (validert.isPresent()) {
            return validert.get();
        }

        Long id = (long) (oppgaver.size() + 1);
        oppgave.put("id", id);

        oppgaver.put(id, oppgave);

        LOG.info("Opprettet oppgave: {}", oppgave);

        return ResponseEntity.status(HttpStatus.CREATED).body(oppgave);
    }

    @GetMapping
    @ApiOperation(value = "Hent oppgaver", response = HentOppgaverResponse.class)
    @ApiImplicitParams({
            @ApiImplicitParam(name = "X-Correlation-ID", required = true, dataType = "string", paramType = "header"),
            @ApiImplicitParam(name = "Authorization", required = true, dataType = "string", paramType = "header"),
            @ApiImplicitParam(name = "tema", dataType = "string", paramType = "query"),
            @ApiImplicitParam(name = "oppgavetype", dataType = "string", paramType = "query"),
            @ApiImplicitParam(name = "journalpostId", dataType = "string", paramType = "query"),
            @ApiImplicitParam(name = "aktoerId", dataType = "string", paramType = "query")
    })
    public ResponseEntity hentOppgaver(
            @Context HttpHeaders httpHeaders,
            @Context UriInfo uriInfo) {
        Optional<ResponseEntity> validert = validerIkkeFunksjonelt(httpHeaders);
        if (validert.isPresent()) {
            return validert.get();
        }

        MultivaluedMap<String, String> queries = uriInfo.getQueryParameters();

        List<ObjectNode> matching = new ArrayList<>();

        oppgaver.forEach((id, oppgave) -> {
            AtomicInteger matches = new AtomicInteger();
            queries.forEach((queryKey, queryValues) -> {
                if (oppgave.hasNonNull(queryKey) && !queryValues.isEmpty()) {
                    JsonNode oppgaveValue = oppgave.get(queryKey);
                    String firstQueryValue = queryValues.get(0);
                    switch (oppgaveValue.getNodeType()) {
                        case STRING:
                            if (oppgaveValue.asText().equals(firstQueryValue)) matches.getAndIncrement();
                            break;
                        case NUMBER:
                            if (oppgaveValue.asLong() == Long.parseLong(firstQueryValue)) matches.getAndIncrement();
                            break;
                        case BOOLEAN:
                            if (oppgaveValue.asBoolean() == firstQueryValue.equals("true")) matches.getAndIncrement();
                            break;
                        default:
                            throw new IllegalStateException("Ikke støttet NodeType " + oppgaveValue.getNodeType().name());
                    }
                }
            });
            if (matches.get() == queries.size()) matching.add(oppgave);
        });

        return ResponseEntity.ok(new HentOppgaverResponse(matching));
    }

    private Optional<ResponseEntity> validerIkkeFunksjonelt(HttpHeaders httpHeaders) {
        // Validerer token på rett format
        String jwt = httpHeaders.getHeaderString("Authorization");
        if (jwt == null || !jwt.startsWith("Bearer")) {
            LOG.error("Ugyldig/manglende Authorization header");
            return Optional.of(ResponseEntity.badRequest().build());

        }
        // Caliderer at correlation id er satt
        String correlationId = httpHeaders.getHeaderString("X-Correlation-ID");
        if (correlationId == null || correlationId.isBlank()) {
            return Optional.of(ResponseEntity.status(HttpStatus.BAD_GATEWAY).build());
        }

        return Optional.empty();
    }
}

File: vtp-pensjon-application/src/main/kotlin/no/nav/pensjon/vtp/mocks/oppgave/rest/OppgaveMockImpl.kt
package no.nav.pensjon.vtp.mocks.oppgave.rest

import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.Parameter
import io.swagger.v3.oas.annotations.Parameters
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@Tag(name = "Oppgave Mock")
@RequestMapping(
    value = ["/rest/oppgave/api/v1/oppgaver"]
)
class OppgaveMockImpl {
    private val oppgaver = mutableListOf<Oppgave>()
    private var oppgaveId = 1L

    private val oppfolginger = mutableListOf<Oppfolging>()
    private var oppfolgingId = 1L

    @PostMapping
    @Operation(summary = "Opprett oppgave")
    @Parameters(
        Parameter(
            name = "X-Correlation-ID",
            required = true,
        ),
        Parameter(name = "Authorization", required = true)
    )
    fun opprettOppgave(
        @RequestBody oppgave: Oppgave,
        @RequestHeader httpHeaders: HttpHeaders?
    ): ResponseEntity<*> {
        oppgave.id = oppgaveId++
        oppgaver.add(oppgave)
        return ResponseEntity.status(HttpStatus.CREATED).body(oppgave)
    }

    @PutMapping("/{oppgaveid}")
    @Operation(summary = "Oppdater oppgave")
    @Parameters(
        Parameter(
            name = "X-Correlation-ID",
            required = true,
        ),
        Parameter(name = "Authorization", required = true)
    )
    fun oppdaterOppgave(
        @PathVariable("oppgaveid") oppgaveId: Long,
        @RequestBody oppgave: Oppgave,
        @RequestHeader httpHeaders: HttpHeaders?
    ): ResponseEntity<*> {
        oppgaver.removeIf { it.id == oppgaveId }
        oppgaver.add(oppgave)
        return ResponseEntity.status(HttpStatus.OK).build<Any>()
    }

    @GetMapping
    @Operation(summary = "Hent oppgaver")
    @Parameters(
        Parameter(
            name = "X-Correlation-ID",
            required = true,
        ),
        Parameter(name = "Authorization", required = true),
        Parameter(name = "tema"),
        Parameter(name = "oppgavetype"),
        Parameter(name = "journalpostId"),
        Parameter(name = "aktoerId")
    )
    fun hentOppgaver(@RequestHeader httpHeaders: HttpHeaders?) =
        HentOppgaverResponse(antallTreffTotalt = oppgaver.size, oppgaver = oppgaver)

    @GetMapping("/{oppgaveid}")
    @Operation(summary = "Hent oppgave")
    @Parameters(
        Parameter(
            name = "X-Correlation-ID",
            required = true,
        ),
        Parameter(name = "Authorization", required = true)
    )
    fun hentOppgave(
        @RequestHeader httpHeaders: HttpHeaders?,
        @PathVariable("oppgaveid") oppgaveId: Long
    ) =
        oppgaver.firstOrNull { it.id == oppgaveId }

    @PatchMapping
    @Operation(summary = "Ferdigstille oppgaver")
    @Parameters(
        Parameter(
            name = "X-Correlation-ID",
            required = true,
        ),
        Parameter(name = "Authorization", required = true)
    )
    fun bulkUpdateOppgaver(): FerdigstillOppgaverResponse {
        val antallOppdaterte = oppgaver.size
        // Pesys benytter dette bare til å ferdigstille oppgaver
        oppgaver.clear()
        oppfolginger.clear()
        return FerdigstillOppgaverResponse(suksess = antallOppdaterte, totalt = antallOppdaterte)
    }

    @GetMapping("/{oppgaveid}/oppfolging")
    @Operation(summary = "Hent oppfølginger for oppgave")
    @Parameters(
        Parameter(
            name = "X-Correlation-ID",
            required = true,
        ),
        Parameter(name = "Authorization", required = true)
    )
    fun hentOppfolginger(
        @PathVariable("oppgaveid") oppgaveId: Long
    ): ResponseEntity<*> {
        val oppfolginger = oppfolginger.filter { it.oppgaveId == oppgaveId }
        return ResponseEntity.status(HttpStatus.OK).body(oppfolginger)
    }

    @PostMapping("/{oppgaveid}/oppfolging")
    @Operation(summary = "Opprett oppfølging")
    @Parameters(
        Parameter(
            name = "X-Correlation-ID",
            required = true,
        ),
        Parameter(name = "Authorization", required = true)
    )
    fun opprettOppfolging(
        @PathVariable("oppgaveid") oppgaveId: Long,
        @RequestBody oppfolging: Oppfolging,
        @RequestHeader httpHeaders: HttpHeaders?
    ): ResponseEntity<*> {
        oppfolging.id = oppfolgingId++
        oppfolginger.add(oppfolging.apply { this.oppgaveId = oppgaveId })
        return ResponseEntity.status(HttpStatus.CREATED).body(oppfolging)
    }

    @PutMapping("/{oppgaveid}/oppfolging/{oppfolgingid}")
    @Operation(summary = "Oppdater oppfølging")
    @Parameters(
        Parameter(
            name = "X-Correlation-ID",
            required = true,
        ),
        Parameter(name = "Authorization", required = true)
    )
    fun oppdaterOppfolging(
        @PathVariable("oppgaveid") oppgaveId: Long,
        @RequestBody oppfolging: Oppfolging,
        @RequestHeader httpHeaders: HttpHeaders?
    ) {
        oppfolginger.removeIf { e -> e.id == oppfolging.id }
        oppfolginger.add(oppfolging.apply { this.oppgaveId = oppgaveId })
    }
}
